Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/business/wrapper/CertLogicEngineWrapper.kt

```diff
@@ -52,24 +52,35 @@ class CertLogicEngineWrapper @Inject constructor(
             valueSetWrapper.valueMap.first()
         )
 
-        Timber.i("Rules to be validated are:")
+        Timber.tag(TAG).i("Rules to be validated are:")
         rules.forEach {
-            Timber.i("Rule ${it.identifier} ${it.version}.")
+            Timber.tag(TAG).i("Rule ${it.identifier} ${it.version}.")
         }
 
-        return engine.validate(
-            hcertVersionString = certificate.certificate.version,
-            rules = rules.map { it.asExternalRule },
-            externalParameter = externalParameter,
-            payload = certificate.certificateJson,
-            certificateType = certificate.asExternalType
-        ).map {
-            it.asEvaluatedDccRule
-        }.toSet().also {
-            Timber.i("Evaluated rules are:")
-            it.forEach {
-                Timber.i("Rule ${it.rule.identifier} ${it.rule.version} has resulted in ${it.result}.")
+        return engine
+            .validate(
+                hcertVersionString = certificate.certificate.version,
+                rules = rules.map { it.asExternalRule },
+                externalParameter = externalParameter,
+                payload = certificate.certificateJson,
+                certificateType = certificate.asExternalType
+            )
+            .map { result ->
+                result.validationErrors?.forEach {
+                    Timber.tag(TAG).e(it, "Errors during validation of %s", result.rule.identifier)
+                }
+                result.asEvaluatedDccRule
             }
-        }
+            .also { evaluated ->
+                Timber.tag(TAG).i("Evaluated rules are:")
+                evaluated.forEach {
+                    Timber.tag(TAG).i("Rule ${it.rule.identifier} ${it.rule.version} has resulted in ${it.result}.")
+                }
+            }
+            .toSet()
```

Comment:
I moved the `toSet()` to the end here so we can see duplicate evaluations in the logs.